src/Common/src/System/Net/SocketAddress.cs

```diff
@@ -229,18 +229,50 @@ public override int GetHashCode()
 
         public override string ToString()
         {
-            var sb = new StringBuilder().Append(Family.ToString()).Append(':').Append(Size).Append(":{");
+            // Get the address family string.  In almost all cases, this should be a cached string
+            // from the enum and won't actually allocate.
+            string familyString = Family.ToString();
 
+            // Determine the maximum length needed to format.
+            int maxLength =
+                familyString.Length + // AddressFamily
+                1 + // :
+                10 + // Size (max length for a positive Int32)
+                2 + // :{
+                (Size - DataOffset) * 4 + // at most ','+3digits per byte
+                1; // }
+
+            Span<char> result = maxLength <= 256 ? // arbitrary limit that should be large enough for the vast majority of cases
+                stackalloc char[256] :
+                new char[maxLength];
+
+            familyString.AsSpan().CopyTo(result);
+            int length = familyString.Length;
+
+            result[length++] = ':';
+
+            bool formatted = Size.TryFormat(result.Slice(length), out int charsWritten);
+            Debug.Assert(formatted);
+            length += charsWritten;
+
+            result[length++] = ':';
+            result[length++] = '{';
+
+            byte[] buffer = Buffer;
             for (int i = DataOffset; i < Size; i++)
             {
                 if (i > DataOffset)
                 {
-                    sb.Append(',');
+                    result[length++] = ',';
```

Comment:
Is it worth it to special case the `i == DataOffset` path outside this loop and remove the condition within the loop?

Reply:
We could measure and change it subsequently.  I was mainly focused here on reducing the allocation in response to a request; even if this slightly reduced measurable CPU overheads, I doubt it would have any real impact.